lib/v8-to-istanbul.js

```diff
@@ -51,7 +52,29 @@ module.exports = class V8ToIstanbul {
     if (this.rawSourceMap) {
       if (this.rawSourceMap.sourcemap.sources.length > 1) {
         this.sourceMap = await new SourceMapConsumer(this.rawSourceMap.sourcemap)
-        this.covSources = this.sourceMap.sourcesContent.map((rawSource, i) => ({ source: new CovSource(rawSource, this.wrapperLength), path: this.sourceMap.sources[i] }))
+        // https://github.com/istanbuljs/v8-to-istanbul/issues/166
+        if (!this.sourceMap.sourcesContent) {
+          this.sourceMap.sourcesContent = []
+          for (const relativePath of this.sourceMap.sources) {
+            const fileDir = path.resolve(this.path, '..')
+            const realPath = path.resolve(fileDir, relativePath)
+            try {
+              const rawSource = await readFile(realPath, 'utf-8')
+              this.sourceMap.sourcesContent.push(rawSource)
+            } catch (err) {
+              // Perhaps some libraries do not provide source files.
+            }
+          }
+        }
+        this.covSources = this.sourceMap.sourcesContent.map((rawSource, i) => {
+          if (!rawSource) {
+            return null
+          }
```

Comment:
is this still needed?

Reply:
The following value may occur: `"sourcesContent": ["xxxx", null],`

Reply:
How about moving the `readFile` behaviour here, instead of read only when the whole `sourcesContent` is falsy?

Reply:
Why? A bundle file may only contain `sources` and not `sourceContent`, then need to read source file. For example, esbuild `build` with `sourceContent` is `false`.

Reply:
@bvanjoi Yes. But I am not suggesting removing the "`= []` when `sourcesContent` is falsy" logic. I am suggesting decoupling it from the `readFile` loop. Currently we read files when and only when `sourcesContent` is falsy, but not read when `sourcesContent` is given and the given `sourcesContent` has some `null` inside.

This isn't necessary. The original issue led me here is actually a bug (https://github.com/evanw/esbuild/issues/1699) in `esbuild`, which may generate a source map with `sourcesContent` given, and the given `sourcesContent` has some `null` inside. But in another view, when a working source-map-provider that has no such bug, provides a such `sourcesContent`, it seems these `null`s should be what it is, and we shouldn't read from somewhere else. So this is an open question. 🤔

Reply:
It would be nice to address the use case of `sourceContent = ["var a = b", null]` since we have seen this crop up in the wild at least once.

What if you default `sourcesContent` to `[]` if it's not defined, and then just do this:

```js
this.covSrouces = []
for (const content of sourcesContent) {
  if (content) {
    this.covSources.push(content);
  } else {
    this.covSources.push(await readFile());
  }
}
```